configure.ac

```diff
@@ -594,20 +594,26 @@ AC_TRY_RUN([
 #ifdef __APPLE__
 #include <mach-o/getsect.h>
 #else
-extern char data_start;
-extern char end;
+#pragma weak __data_start
+#pragma weak _end
+extern int __data_start;
+extern int _end;
 #endif
 
+int data;
+
 int main(void) {
     void *base;
-    unsigned long length;
+    long length;
 #ifdef __APPLE__
     base = (void*) get_etext();
     length = get_end() - get_etext();
 #else
-    base = &data_start;
-    length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == (int*) 0 || &_end == (int*) 0) return 1;
+    base = (void*) &__data_start;
+    length = (long) ((char*) &_end - (char*) &__data_start);
```

Comment:
The `(char*)` cast is slightly perplexing to me, but it looks like it works out...  Do we need it?  E.g., would `(long)&end - (long)&__data_start` not work?

Reply:
Casting a pointer to an integer is bad (unless it's intptr_t); it's better to do the arithmetic on the pointer type.  If we do the arithmetic on int*, we need to multiply the result by sizeof(int).  Casting to char* tells us directly how many bytes, which is the number we need.

Reply:
Ah yes, agreed.